edx-solutions/xblock-drag-and-drop-v2
# [SOL-1909] Multiple Drop Targets per Item

This PR implements SOL-1909, and allows instructors to select multiple valid options for each item in a Drag and Drop v2 problem. It is fully backwards-compatible with existing problem sets.

We approach this by changing the `zone` attribute on each item to a `zones` attribute containing a list, which can contain an arbitrary number of zone IDs.  Additionally, a compatibility layer has been introduced such that existing datasets with `zone` are mutated before being passed to the frontend to have a list containing the existing single item (in cases where there's an ID value) or an empty list (in cases where `zone` is set to `"none"`). This won't migrate existing datasets automatically, but they will be updated the next time the XBlock configuration is saved. 

**JIRA tickets**: Implements [SOL-1909](https://openedx.atlassian.net/browse/SOL-1909)

**Dependencies**: None

**Screenshots**:

<img width="1346" alt="screen shot 2016-07-22 at 11 24 47 am" src="https://cloud.githubusercontent.com/assets/7773758/17063996/c704ed1e-4fff-11e6-89e7-f07d18a9d21b.png">

<img width="1348" alt="screen shot 2016-07-22 at 11 24 35 am" src="https://cloud.githubusercontent.com/assets/7773758/17063995/c702ec44-4fff-11e6-8ae9-3135d5d0052f.png">

![screen shot 2016-07-20 at 2 07 35 pm](https://cloud.githubusercontent.com/assets/7773758/16997436/7c93ebcc-4e83-11e6-940e-9dc9d312cb3c.png)

![screen shot 2016-07-20 at 2 07 23 pm](https://cloud.githubusercontent.com/assets/7773758/16997439/7ca1aed8-4e83-11e6-8daf-557ad615150d.png)

![screen shot 2016-07-20 at 2 07 08 pm](https://cloud.githubusercontent.com/assets/7773758/16997437/7c9e9f18-4e83-11e6-8234-95f2bb98872a.png)

**Sandbox URL**:

- LMS: http://dndv2-sandbox4.opencraft.hosting/

- Studio: http://studio-dndv2-sandbox4.opencraft.hosting/

**Testing instructions**:

1. In Studio, create a course or update a course to include the DnDv2 XBlock

2. In Studio, observe that the dropdown select menu for Zone selection has changed to be multiple checkbox controls.

3. In Studio, observe that existing configurations are retained and presented appropriately for the new control set.

4. In Studio, create a new item that can be assigned to multiple zones.

5. In LMS, observe that pre-existing items have appropriate behavior when placed on the zone they were previously assigned to.

6. In LMS, observe that the item you created can be dragged to any of the zones you selected for it.

7. In LMS, review the XBlock demo course and observe that it works without modification, and that changes to take advantage of the feature allowing multiple correct answers works immediately.

**Reviewers**
- [x] a11y: @cptvitamin and/or @clrux

Also tagging @pdesjardins as a docs FYI.
